File: CodeCast-desktop/scripts/generate-settings-keys.mjs
/**
 * generate-settings-keys.mjs
 *
 * 从 Go Settings struct 的 json tag 自动生成 TypeScript 常量文件
 * 用法: node scripts/generate-settings-keys.mjs
 */

import { readFileSync, writeFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const rootDir = join(__dirname, '..');
const srcPath = join(rootDir, 'main.go');
const outPath = join(rootDir, 'frontend', 'src', 'settingsKeys.ts');

const src = readFileSync(srcPath, 'utf-8');

const jsonTagRe = /`json:"([^"]+)"`/;
const keys = [];
const arrayKeys = [];
let inSettings = false;

const arrayTypePatterns = ['[]MCPServer', '[]EnvVar', '[]SlashCommand', '[]string'];

for (const line of src.split('\n')) {
  if (line.includes('type Settings struct')) {
    inSettings = true;
    continue;
  }
  if (inSettings && line.trim() === '}') break;
  if (!inSettings) continue;

  const match = line.match(jsonTagRe);
  if (!match) continue;

  const key = match[1];
  if (!key || key.includes(',')) continue;

  const isArray = arrayTypePatterns.some(p => line.includes(p));
  if (isArray) {
    arrayKeys.push(key);
  } else {
    keys.push(key);
  }
}

const categorized = {};
for (const k of keys) {
  let cat = 'other';
  if (/work_mode|default_perm|auto_review|full_access|shell|open_target|language|hotkey|ctrl_enter|followup|review_mode/.test(k)) cat = 'work';
  else if (/notify|notification/.test(k)) cat = 'notify';
  else if (/theme|font_size/.test(k)) cat = 'ui';
  else if (/api_key|long_context/.test(k)) cat = 'api';
  else if (/personality|custom_instructions|auto_memory|tool_memory|message_history/.test(k)) cat = 'memory';
  else if (/commit|worktree/.test(k)) cat = 'git';
  else if (/browser|domain|selenium|allow_browser/.test(k)) cat = 'browser';
  else if (/computer_control/.test(k)) cat = 'computer';

  (categorized[cat] ??= []).push(k);
}

const catLabels = {
  work: '工作模式',
  notify: '通知',
  ui: '外观',
  api: 'API',
  memory: '个性化 / 记忆',
  git: 'Git',
  browser: '浏览器',
  computer: '计算机控制',
};

const catOrder = ['work', 'notify', 'ui', 'api', 'memory', 'git', 'browser', 'computer'];

let output = `/**
 * settingsKeys.ts — 由 scripts/generate-settings-keys.mjs 自动生成
 *
 * ⚠️ 请勿手动编辑此文件！修改后端 Settings struct 后重新运行:
 *   node scripts/generate-settings-keys.mjs
 *
 * 生成时间: ${new Date().toISOString()}
 */

export const S = {
`;

for (const cat of catOrder) {
  const items = categorized[cat];
  if (!items?.length) continue;
  output += `\n  // ===== ${catLabels[cat]} =====\n`;
  for (const k of items) {
    output += `  ${k}: '${k}',\n`;
  }
}

if (arrayKeys.length > 0) {
  output += `\n  // ===== 数组 / 对象类型（通过完整 API 操作） =====\n`;
  for (const k of arrayKeys) {
    output += `  ${k}: '${k}',\n`;
  }
}

output += `} as const;

export type SettingKey = typeof S[keyof typeof S];

/** 所有可设置的标量 key（用于 toggle/select/input 绑定） */
export const SCALAR_KEYS: readonly SettingKey[] = [
`;
for (const k of keys) {
  output += `  S.${k},\n`;
}
output += '] as const;\n';

writeFileSync(outPath, output, 'utf-8');
console.log(`✅ 已生成 ${outPath}`);
console.log(`   标量字段: ${keys.length} 个`);
if (arrayKeys.length > 0) console.log(`   数组字段: ${arrayKeys.length} 个 (${arrayKeys.join(', ')})`);

File: CodeCast-desktop/scripts/generate-settings-keys.go
//go:build ignore

package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"
)

func main() {
	if len(os.Args) < 3 {
		fmt.Println("用法: go run generate-settings-keys.go <main.go路径> <输出ts路径>")
		fmt.Println("示例: go run generate-settings-keys.go ../main.go settingsKeys.ts")
		os.Exit(1)
	}

	srcPath := os.Args[1]
	outPath := os.Args[2]

	data, err := os.ReadFile(srcPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "读取源文件失败: %v\n", err)
		os.Exit(1)
	}

	re := regexp.MustCompile("`json:\"([^\"]+)\"`")
	inSettings := false
	var keys []string
	skipTypes := map[string]bool{
		"MCPServer":   true,
		"EnvVar":      true,
		"SlashCommand": true,
	}

	scanner := bufio.NewScanner(strings.NewReader(string(data)))
	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "type Settings struct") {
			inSettings = true
			continue
		}
		if inSettings && strings.HasPrefix(line, "}") {
			break
		}
		if !inSettings {
			continue
		}
		for _, skipType := range []string{"[]MCPServer", "[]EnvVar", "[]SlashCommand", "[]string"} {
			if strings.Contains(line, skipType) {
				inSettings = false
				break
			}
		}
		matches := re.FindStringSubmatch(line)
		if len(matches) >= 2 {
			key := matches[1]
			if key != "" && !strings.Contains(key, ",") && !skipTypes[key] {
				keys = append(keys, key)
			}
		}
	}

	sort.Strings(keys)

	var sb strings.Builder
	sb.WriteString(`/**
 * settingsKeys.ts — 由 generate-settings-keys.go 自动生成
 *
 * ⚠️ 请勿手动编辑此文件！修改后端 Settings struct 后重新运行:
 *   go run scripts/generate-settings-keys.go main.go frontend/src/settingsKeys.ts
 */

export const S = {
`)

	categories := map[string][]string{
		"工作模式":    {},
		"通知":       {},
		"外观":       {},
		"API":        {},
		"个性化/记忆": {},
		"Git":        {},
		"浏览器":     {},
		"计算机控制":  {},
	}

	modeCat := ""
	for _, k := range keys {
		switch {
		case strings.Contains(k, "work_mode") || strings.Contains(k, "perm") || strings.Contains(k, "review") || strings.Contains(k, "access") || strings.Contains(k, "shell") || strings.Contains(k, "open_target") || strings.Contains(k, "language") || strings.Contains(k, "hotkey") || strings.Contains(k, "ctrl_enter") || strings.Contains(k, "followup") || strings.Contains(k, "review_mode"):
			categories["工作模式"] = append(categories["工作模式"], k)
		case strings.Contains(k, "notify") || strings.Contains(k, "notification"):
			categories["通知"] = append(categories["通知"], k)
		case strings.Contains(k, "theme") || strings.Contains(k, "font_size"):
			categories["外观"] = append(categories["外观"], k)
		case strings.Contains(k, "api_key") || strings.Contains(k, "long_context"):
			categories["API"] = append(categories["API"], k)
		case strings.Contains(k, "personality") || strings.Contains(k, "custom_instructions") || strings.Contains(k, "auto_memory") || strings.Contains(k, "tool_memory") || strings.Contains(k, "message_history"):
			categories["个性化/记忆"] = append(categories["个性化/记忆"], k)
		case strings.Contains(k, "commit") || strings.Contains(k, "worktree"):
			categories["Git"] = append(categories["Git"], k)
		case strings.Contains(k, "browser") || strings.Contains(k, "domain") || strings.Contains(k, "selenium"):
			categories["浏览器"] = append(categories["浏览器"], k)
		case strings.Contains(k, "computer_control"):
			categories["计算机控制"] = append(categories["计算机控制"], k)
		default:
			categories["工作模式"] = append(categories["工作模式"], k)
		}
	}

	catOrder := []string{"工作模式", "通知", "外观", "API", "个性化/记忆", "Git", "浏览器", "计算机控制"}
	for _, cat := range catOrder {
		if len(categories[cat]) == 0 {
			continue
		}
		sb.WriteString(fmt.Sprintf("\n  // ===== %s =====\n", cat))
		for _, k := range categories[cat] {
			sb.WriteString(fmt.Sprintf("  %s: '%s',\n", k, k))
		}
	}

	sb.WriteString(`} as const;

export type SettingKey = typeof S[keyof typeof S];

/** 所有可设置的标量 key（用于 toggle/select/input 绑定） */
export const SCALAR_KEYS: readonly SettingKey[] = [
`)
	for _, k := range keys {
		if !strings.Contains(k, "servers") && !strings.Contains(k, "vars") && !strings.Contains(k, "commands") && !strings.Contains(k, "sessions") {
			sb.WriteString(fmt.Sprintf("  S.%s,\n", k))
		}
	}
	sb.WriteString("] as const;\n")

	if err := os.WriteFile(outPath, []byte(sb.String()), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "写入失败: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("✅ 已生成 %s (%d 个字段)\n", outPath, len(keys))
}
